packages/core/integration-tests/test/scope-hoisting.js

```diff
@@ -1964,6 +1964,31 @@ describe('scope hoisting', function () {
       assert(new output[3]() instanceof output[2]);
     });
 
+    it("considers an asset's closest package.json for sideEffects, not the package through which it found the asset", async () => {
+      let b = await bundle(
+        path.join(
+          __dirname,
+          '/integration/scope-hoisting/es6/side-effects-package-redirect/index.js',
+        ),
+      );
+
+      let result = await run(b);
+      assert.strictEqual(result, 1);
+
+      let bar;
+      b.traverse((node, _, actions) => {
+        if (
+          node.type === 'asset' &&
+          node.value.filePath.endsWith('real-bar.js')
+        ) {
+          bar = node.value;
+          actions.stop();
+        }
+      });
```

Comment:
We have a test utility for this: https://github.com/parcel-bundler/parcel/blob/a8a4f5114944512aad60df9baffa741f31705c8d/packages/core/test-utils/src/utils.js#L143-L146